File: indic-script-converter-js/tests/deva-gujr-converter.test.js
import assert from 'node:assert/strict';
import test from 'node:test';

import { devanagariSmokeSamples } from '../src/example-deva.js';
import { gujaratiSmokeSamples } from '../src/example-gujr.js';
import {
    IndicScriptConversionOptions,
    toCanonicalDevanagariFromGujarati,
    toCanonicalGujaratiFromDevanagari,
    toDevanagariFromGujarati,
    toExactDevanagariFromGujarati,
    toExactGujaratiFromDevanagari,
    toGujaratiFromDevanagari,
} from '../src/deva-gujr-converter.js';
import { IastToGujaratiOptions, toGujaratiFromIast } from '../src/latn-iast-to-gujr.js';

test('core canonical script conversion', () => {
    assert.equal(toCanonicalGujaratiFromDevanagari('कृष्ण'), 'કૃષ્ણ');
    assert.equal(toCanonicalDevanagariFromGujarati('કૃષ્ણ'), 'कृष्ण');
    assert.equal(toCanonicalGujaratiFromDevanagari('१२३'), '૧૨૩');
    assert.equal(toCanonicalDevanagariFromGujarati('૧૨૩'), '१२३');
    assert.equal(toCanonicalGujaratiFromDevanagari('वसोः॑'), 'વસોઃ॑');
    assert.equal(toCanonicalDevanagariFromGujarati('વસોઃ॑'), 'वसोः॑');
});

test('nukta and extended mappings are consumed as whole tokens', () => {
    assert.equal(toCanonicalGujaratiFromDevanagari('क़ ख़ ग़ ज़ फ़'), 'ક઼ ખ઼ ગ઼ જ઼ ફ઼');
    assert.equal(toCanonicalDevanagariFromGujarati('ક઼ ખ઼ ગ઼ જ઼ ફ઼'), 'क़ ख़ ग़ ज़ फ़');
    assert.equal(toCanonicalGujaratiFromDevanagari('ॿक्ति'), 'બ઼ક્તિ');
    assert.equal(toCanonicalDevanagariFromGujarati('બ઼ક્તિ'), 'ॿक्ति');
    assert.equal(toCanonicalGujaratiFromDevanagari('ॹ'), 'ૹ');
    assert.equal(toCanonicalDevanagariFromGujarati('ૹ'), 'ॹ');
});

test('exact metadata round trips every Devanagari corpus item', () => {
    const options = new IndicScriptConversionOptions({
        embedExactSourceMetadata: true,
    });
    for (const source of devanagariSmokeSamples) {
        const taggedGujarati = toCanonicalGujaratiFromDevanagari(source, options);
        assert.equal(toExactDevanagariFromGujarati(taggedGujarati), source);
        assert.equal(toDevanagariFromGujarati(taggedGujarati), source);
    }
});

test('exact metadata round trips every Gujarati corpus item', () => {
    const options = new IndicScriptConversionOptions({
        embedExactSourceMetadata: true,
    });
    for (const source of gujaratiSmokeSamples) {
        const taggedDevanagari = toCanonicalDevanagariFromGujarati(source, options);
        assert.equal(toExactGujaratiFromDevanagari(taggedDevanagari), source);
        assert.equal(toGujaratiFromDevanagari(taggedDevanagari), source);
    }
});

test('visible-script tampering invalidates exact recovery', () => {
    const options = new IndicScriptConversionOptions({
        embedExactSourceMetadata: true,
    });
    const tagged = toCanonicalGujaratiFromDevanagari('कृष्ण', options);
    const tampered = tagged.replace('કૃષ્ણ', 'રામ');
    assert.throws(() => toExactDevanagariFromGujarati(tampered), TypeError);
});

test('typed metadata rejects unrelated Latin-source trailers', () => {
    const taggedFromLatin = toGujaratiFromIast(
        'Kṛṣṇa',
        new IastToGujaratiOptions({ embedExactSourceMetadata: true }),
    );
    assert.throws(() => toExactDevanagariFromGujarati(taggedFromLatin), TypeError);
});

File: dart/test/deva_gujr_converter_test.dart
import 'package:lipimala/deva_gujr_converter.dart';
import 'package:lipimala/latn_iast_to_gujr.dart';
import 'package:test/test.dart';

import '../tool/latn_iast_transliteration_verification/example_deva.dart'
    as devaExamples;
import '../tool/latn_iast_transliteration_verification/example_gujr.dart'
    as gujrExamples;

void main() {
  test('core canonical script conversion', () {
    expect('कृष्ण'.toCanonicalGujaratiFromDevanagari(), 'કૃષ્ણ');
    expect('કૃષ્ણ'.toCanonicalDevanagariFromGujarati(), 'कृष्ण');
    expect('१२३'.toCanonicalGujaratiFromDevanagari(), '૧૨૩');
    expect('૧૨૩'.toCanonicalDevanagariFromGujarati(), '१२३');
    expect('वसोः॑'.toCanonicalGujaratiFromDevanagari(), 'વસોઃ॑');
    expect('વસોઃ॑'.toCanonicalDevanagariFromGujarati(), 'वसोः॑');
  });

  test('nukta and extended mappings are whole tokens', () {
    expect(
      'क़ ख़ ग़ ज़ फ़'.toCanonicalGujaratiFromDevanagari(),
      'ક઼ ખ઼ ગ઼ જ઼ ફ઼',
    );
    expect(
      'ક઼ ખ઼ ગ઼ જ઼ ફ઼'.toCanonicalDevanagariFromGujarati(),
      'क़ ख़ ग़ ज़ फ़',
    );
    expect('ॿक्ति'.toCanonicalGujaratiFromDevanagari(), 'બ઼ક્તિ');
    expect('બ઼ક્તિ'.toCanonicalDevanagariFromGujarati(), 'ॿक्ति');
    expect('ॹ'.toCanonicalGujaratiFromDevanagari(), 'ૹ');
    expect('ૹ'.toCanonicalDevanagariFromGujarati(), 'ॹ');
  });

  test('exact metadata round trips every Devanagari corpus item', () {
    const options = IndicScriptConversionOptions(
      embedExactSourceMetadata: true,
    );
    for (final source in devaExamples.devanagariSmokeSamples) {
      final tagged = source.toCanonicalGujaratiFromDevanagari(
        options: options,
      );
      expect(tagged.toExactDevanagariFromGujarati(), source);
      expect(tagged.toDevanagariFromGujarati(), source);
    }
  });

  test('exact metadata round trips every Gujarati corpus item', () {
    const options = IndicScriptConversionOptions(
      embedExactSourceMetadata: true,
    );
    for (final source in gujrExamples.gujaratiSmokeSamples) {
      final tagged = source.toCanonicalDevanagariFromGujarati(
        options: options,
      );
      expect(tagged.toExactGujaratiFromDevanagari(), source);
      expect(tagged.toGujaratiFromDevanagari(), source);
    }
  });

  test('visible tampering invalidates exact metadata', () {
    const options = IndicScriptConversionOptions(
      embedExactSourceMetadata: true,
    );
    final tagged = 'कृष्ण'.toCanonicalGujaratiFromDevanagari(
      options: options,
    );
    final tampered = tagged.replaceFirst('કૃષ્ણ', 'રામ');
    expect(
      tampered.toExactDevanagariFromGujarati,
      throwsFormatException,
    );
  });

  test('typed metadata rejects unrelated Latin-source trailers', () {
    const options = IastToGujaratiOptions(
      embedExactSourceMetadata: true,
    );
    final taggedFromLatin = 'Kṛṣṇa'.toGujaratiFromIast(options: options);
    expect(
      taggedFromLatin.toExactDevanagariFromGujarati,
      throwsFormatException,
    );
  });
}
